src/main/java/com/amazon/opendistroforelasticsearch/knn/index/util/KNNLibrary.java

```diff
@@ -0,0 +1,214 @@
+            this.latestLibraryVersion = latestLibraryVersion;
+            this.extension = extension;
+        }
+
+        @Override
+        public String getLatestBuildVersion() {
+            return this.latestLibraryBuildVersion;
+        }
+
+        @Override
+        public String getLatestLibVersion() {
+            return this.latestLibraryVersion;
+        }
+
+        @Override
+        public String getExtension() {
+            return this.extension;
+        }
+
+        @Override
+        public String getCompoundExtension() {
+            return getExtension() + "c";
+        }
+
+        @Override
+        public KNNMethod getMethod(String methodName) {
+            if (!methods.containsKey(methodName)) {
+                throw new IllegalArgumentException("Invalid method name: " + methodName);
+            }
+            return methods.get(methodName);
+        }
+
+        @Override
+        public float score(float rawScore, SpaceType spaceType) {
+            if (this.scoreTranslation.containsKey(spaceType)) {
+                return this.scoreTranslation.get(spaceType).apply(rawScore);
+            }
+
+            return spaceType.scoreTranslation(rawScore);
+        }
+
+        @Override
+        public void validateMethod(KNNMethodContext knnMethodContext) {
+            String methodName = knnMethodContext.getMethodComponent().getName();
+            if (!methods.containsKey(methodName)) {
+                throw new ValidationException();
+            }
+
+            KNNMethod knnMethod = methods.get(methodName);
+            knnMethod.validate(knnMethodContext);
+        }
+    }
+
+    /**
+     * Implements NativeLibrary for the nmslib native library
+     */
+    class Nmslib extends NativeLibrary {
+        // ======================================
+        // Constants pertaining to nmslib library
+        // ======================================
+        public final static String HNSW_LIB_NAME = "hnsw";
+        public final static String EXTENSION = ".hnsw";
+
+        public final static Map<String, KNNMethod> METHODS = ImmutableMap.of(
+                METHOD_HNSW,
+                KNNMethod.Builder.builder(
+                        MethodComponent.Builder.builder(HNSW_LIB_NAME)
+                                .addParameter(METHOD_PARAMETER_M, new Parameter.IntegerParameter(16,
+                                        v -> v > 0))
+                                .addParameter(METHOD_PARAMETER_EF_CONSTRUCTION, new Parameter.IntegerParameter(512,
+                                        v -> v > 0))
+                                .build())
+                        .addSpaces(SpaceType.L2, SpaceType.L1, SpaceType.LINF, SpaceType.COSINESIMIL,
+                                SpaceType.INNER_PRODUCT)
+                        .build()
+        );
+
+        public final static Map<SpaceType, Function<Float, Float>> SCORE_TRANSLATIONS = Collections.emptyMap();
```

Comment:
nit: do we need this variable? is it used other than line no: 197

Reply:
No, not really.